### ISSUE TITLE
Rule proposal: require-await-async-fn

### ISSUE THREAD

**Issue by silverwind** [sindresorhus/eslint-plugin-unicorn#1144]
Proposing a simple rule to require `await` when the callee is `async`.

## Fail

```js
const a = async () => {};
a(); /* error: missing await on async function call */
```

## Pass

```js
const a = async () => {};
await a();
```

## Notes

There are rare cases where one does not want to await a async function, for example by defering the `await`:

```js
const a = async () => {};
const promise = a();
await promise;
```

Ideally such a pattern would also pass the rule but I could live with having to put a `eslint-disable` comment there.

eslint's [require-await](https://eslint.org/docs/rules/require-await) is insufficient as it does not verify whether the callee is actually `async` and also does not work with top-level `await`.

**Comment by sindresorhus** [comment]
> eslint's require-await is insufficient as it does not verify whether the callee is actually async and also does not work with top-level await.

We can only verify this for some simple cases though.

**Comment by fisker** [comment]
Is this what are you looking for ? https://www.npmjs.com/package/eslint-plugin-no-floating-promise

**Comment by silverwind** [comment]
> Is this what are you looking for ? https://www.npmjs.com/package/eslint-plugin-no-floating-promise

Yes, exactly. I will try this.

**Comment by joealden** [comment]
> Seems to work somewhat but it's very limited ([one issue](https://github.com/SebastienGllmt/eslint-plugin-no-floating-promise/issues/8)).

@silverwind perhaps [`@typescript-eslint/eslint-plugin/no-floating-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md) would not suffer from such problems as easily, as it has access to types for linting? Of course this is only really viable if you have a TS codebase.
